File: c/3-pid/main.c
#include <errno.h>
#include <poll.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>

#include <pigpio.h>
#include <string.h>

#include "controller.h"
#include "registers.h"
#include "server.h"

#define N_FDS_SYSTEM 2
#define N_CONNECTIONS 5
#define N_FDS_MAX (N_FDS_SYSTEM + N_CONNECTIONS)

static const server_options_t SERVER_OPTIONS = {.n_connections = N_CONNECTIONS};

static const uint64_t READ_FREQUENCY = 1000;
static const uint64_t CONTROL_FREQUENCY = 10;
static const uint64_t READS_PER_BIN = (READ_FREQUENCY / CONTROL_FREQUENCY);

static const controller_options_t CONTROLLER_OPTIONS = {
    .control_frequency = CONTROL_FREQUENCY,
    .time_window_bins = 10,
    .reads_per_bin = READS_PER_BIN,
    .revolution_treshold_close = 0.20,
    .revolution_treshold_far = 0.36,
    .pwm_channel = 13,
    .pwm_frequency = 1000.,
};

static bool do_continue = true;

void interrupt_handler(int) {
  printf("\nGracefully stopping\n");
  do_continue = false;
}

int main(int, char **) {
  int res;

  printf("Controlling motor using PID from C\n");

  res = gpioInitialise();
  if (res < 0) {
    fprintf(stderr, "gpioInitialise fail (%d)\n", res);
    return EXIT_FAILURE;
  }

  res = gpioSetSignalFunc(SIGINT, &interrupt_handler);
  if (res < 0) {
    fprintf(stderr, "gpioSetSignalFunc fail (%d)\n", res);
    gpioTerminate();
    return EXIT_FAILURE;
  }

  modbus_mapping_t *registers = registers_init();
  if (registers == NULL) {
    fprintf(stderr, "registers_init fail\n");
    gpioTerminate();
    return EXIT_FAILURE;
  }

  static server_t server;
  res = server_init(&server, registers, SERVER_OPTIONS);
  if (res < 0) {
    fprintf(stderr, "server_init fail (%d)\n", res);
    registers_free(registers);
    gpioTerminate();
    return EXIT_FAILURE;
  }

  static controller_t controller;
  res = controller_init(&controller, registers, CONTROLLER_OPTIONS);
  if (res < 0) {
    fprintf(stderr, "controller_init fail (%d)\n", res);
    server_deinit(&server);
    registers_free(registers);
    gpioTerminate();
    return EXIT_FAILURE;
  }

  struct pollfd poll_fds[N_FDS_MAX] = {
      {.fd = controller.timer_fd, .events = POLL_IN},
      {.fd = server.socket_fd, .events = POLL_IN},
  };
  size_t n_poll_fds = N_FDS_SYSTEM;

  while (do_continue) {
    res = poll(poll_fds, n_poll_fds, 1000);
    if (res == -1 && errno != EINTR)
      fprintf(stderr, "poll fail (%d): %s\n", res, strerror(errno));
    if (res <= 0)
      continue;

    for (size_t i = 0; i < n_poll_fds; ++i) {
      struct pollfd *poll_fd = &poll_fds[i];
      int fd = poll_fd->fd;

      if (poll_fd->revents & (POLLERR | POLLHUP)) {
        poll_fd->fd = -fd; // mark for removal
        server_close_fd(&server, fd);
      }
      if (poll_fd->revents & POLLERR)
        fprintf(stderr, "File (socket?) closed unexpectedly\n");
      if (poll_fd->revents & POLLNVAL)
        fprintf(stderr, "File (socket?) not open\n");
      if (poll_fd->revents & POLLIN) {
        res = controller_handle(&controller, fd);
        if (res < 0) {
          fprintf(stderr, "controller_handle fail (%d)\n", res);
        }
        if (res != 0)
          continue; // Handled -- either error or success

        server_result_t result;
        res = server_handle(&server, fd, &result);
        if (res != 0) {
          fprintf(stderr, "server_handle fail (%d)\n", res);
        }

        // Reflect connection modifications in poll_fds
        if (result.is_closed)
          poll_fd->fd = -fd; // mark for removal
        if (result.new_connection_fd != -1) {
          poll_fds[n_poll_fds++] =
              (struct pollfd){.fd = result.new_connection_fd, .events = POLLIN};
        }
      }
    }

    // Remove marked connections
    size_t i = 0;
    while (i < n_poll_fds) {
      if (poll_fds[i].fd < 0)
        poll_fds[i] = poll_fds[--n_poll_fds];
      else
        i += 1;
    }
  }

  controller_deinit(&controller);
  server_deinit(&server);
  registers_free(registers);
  gpioTerminate();
  return EXIT_SUCCESS;
}

File: zig/3-pid/main.zig
const std = @import("std");
const config = @import("config");
const posix = std.posix;
const POLL = std.os.linux.POLL;

const pwm = @import("pwm");

const Server = @import("Server.zig");
const Registers = @import("Registers.zig");
const Controller = @import("Controller.zig");
const memory = @import("memory.zig");
const c = @import("c.zig");

const n_fds_system = 2;
const n_connections_max = 5;
const n_fds = n_fds_system + n_connections_max;

const i2c_adapter_number = 1;
const i2c_adapter_path = std.fmt.comptimePrint("/dev/i2c-{}", .{i2c_adapter_number});
const ads7830_address = 0x48;

const motor_pwm_channel = 1; // gpio 13
const pwm_frequency: f32 = 1000;

const read_frequency: u32 = 1000;
const control_frequency: u32 = 10;
const reads_per_bin: u32 = read_frequency / control_frequency;

const controller_options: Controller.Options = .{
    .control_frequency = control_frequency,
    .time_window_bins = 10,
    .reads_per_bin = reads_per_bin,
    .revolution_threshold_close = 0.20,
    .revolution_threshold_far = 0.36,
    .i2c_adapter_path = i2c_adapter_path,
    .i2c_address = ads7830_address,
    .pwm_channel = motor_pwm_channel,
    .pwm_frequency = 1000,
};

var do_continue = true;
pub fn interruptHandler(_: c_int) callconv(.C) void {
    std.debug.print("\nGracefully stopping\n", .{});
    do_continue = false;
}

pub fn main() !void {
    std.log.info("Controlling motor using PID from Zig", .{});

    const signal = @intFromPtr(c.signal(c.SIGINT, &interruptHandler));
    if (signal < 0)
        return std.posix.unexpectedErrno(std.posix.errno(signal));

    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    var counting = memory.CountingAllocator.init(gpa.allocator());
    const alloc = counting.allocator();

    var registers = try Registers.init();
    defer registers.deinit();

    var controller = try Controller.init(alloc, &registers, controller_options);
    defer controller.deinit();

    var server = try Server.init(
        alloc,
        &registers,
        .{ .n_connections_max = 5 },
    );
    defer server.deinit();

    var poll_fds = try std.BoundedArray(posix.pollfd, n_fds).init(0);
    const initial_fds = try poll_fds.addManyAsArray(n_fds_system);
    initial_fds.* = .{
        pollfdInit(server.socket.handle),
        pollfdInit(controller.timer.handle),
    };

    while (do_continue) {
        const n_polled = posix.poll(poll_fds.slice(), 1000) catch |err| {
            std.log.err("poll fail: {}", .{err});
            continue;
        };
        if (n_polled == 0)
            continue;

        for (poll_fds.slice()) |*poll_fd| {
            const fd = poll_fd.fd;

            if (poll_fd.revents & POLL.ERR != 0)
                std.log.err("File (socket?) closed unexpectedly", .{});
            if (poll_fd.revents & POLL.NVAL != 0)
                std.log.err("File (socket?) not open", .{});

            if (poll_fd.revents & (POLL.ERR | POLL.HUP) != 0) {
                poll_fd.fd = -fd; // mark for removal
                server.closeConnection(fd);
            } else if (poll_fd.revents & POLL.IN != 0) {
                const controller_res = controller.handle(fd) catch |err| {
                    std.log.err("Controller.handle fail: {}", .{err});
                    continue;
                };
                if (controller_res == .handled)
                    continue;

                const server_res = server.handle(fd) catch |err| {
                    std.log.err("Server.handle fail: {}", .{err});
                    continue;
                };

                switch (server_res) {
                    .accepted => |file| {
                        const new_pollfd = poll_fds.addOneAssumeCapacity();
                        new_pollfd.* = pollfdInit(file.handle);
                    },
                    .closed => poll_fd.fd = -fd, // mark for removal
                    else => {},
                }
            }
        }

        // Remove marked connections
        var i: u32 = 0;
        while (i < poll_fds.len) {
            const poll_fd = &poll_fds.slice()[i];
            if (poll_fd.fd < 0) {
                std.log.info("Closing connection on socket {}", .{-poll_fd.fd});
                poll_fd.* = poll_fds.pop();
            } else i += 1;
        }
    }
}

fn pollfdInit(fd: posix.fd_t) posix.pollfd {
    return .{ .fd = fd, .events = POLL.IN, .revents = 0 };
}

pub const std_options: std.Options = .{
    .log_level = std.enums.nameCast(std.log.Level, config.log_level),
};
